File: 2023/day11.py
import os
import sys
import itertools

test = """\
...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#....."""

day = os.path.splitext(os.path.basename(__file__))[0]

if 'test' in sys.argv:
    data = test.splitlines()
else:
    data = open(day+'.txt').read().splitlines()

DEBUG = 'debug' in sys.argv

# Find the rows without galaxies.

nogrow = [n for n,row in enumerate(data) if '#' not in row]

# Find the columns without galaxies.

counts = [0]*len(data[0])
for row in data:
    for n,col in enumerate(row):
        if col == '#':
            counts[n] += 1

nogcol = [n for n,val in enumerate(counts) if not val]

def expand(data, delta):
    stars = []
    dy = 0
    for y,row in enumerate(data):
        if y in nogrow:
            dy += delta
            continue
        dx = 0
        for x,col in enumerate(row):
            if x in nogcol:
                dx += delta
            elif col == '#':
                stars.append( (x+dx,y+dy))
    return stars

def part1(data,delta):
    stars = expand(data,delta-1)
    mandist = 0
    for i in range(len(stars)-1):
        for j in range(i,len(stars)):
            mandist += abs(stars[i][0]-stars[j][0]) + abs(stars[i][1]-stars[j][1])
    return mandist

print("Part 1:", part1(data,2))
if 'test' in sys.argv:
    print("Test 10:", part1(data,10))
    print("Test 100:", part1(data,100))
print("Part 2:", part1(data,1000000))

File: 2023/day11.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cstring>
#include <vector>
#include <map>
#include <deque>
#include <tuple>
#include <algorithm>
#include <numeric>

using namespace std;

bool DEBUG = false;
bool TEST = false;

typedef vector<string> StringVector;
typedef vector<int> IntVector;
typedef vector<int64_t> LongVector;

const string test(
"...#......\n"
".......#..\n"
"#.........\n"
"..........\n"
"......#...\n"
".#........\n"
".........#\n"
"..........\n"
".......#..\n"
"#...#....."
);


struct Point {
    int x;
    int y;

    bool operator==(const Point & other)
    {
        return x==other.x & y == other.y;
    }
};

bool operator<(const Point & a, const Point & b )
{
    if( a.x == b.x )
        return a.y < b.y;
    return a.x < b.x;
}



void expand( StringVector & lines, int delta, vector<Point> & stars )
{

    // Find the rows without galaxies.

    vector<int> nogrow( lines.size() );
    for( int y = 0; y < lines.size(); y++ )
        if( lines[y].find('#') == string::npos )
            nogrow[y] = 1;

    // Find the columns without galaxies.

    vector<int> counts( lines.front().size() );
    for( auto & row : lines )
        for( int x = 0; x < row.size(); x++ )
            if( row[x] == '#' )
                counts[x] += 1;

    vector<int> nogcol( counts.size() );
    for( int x = 0; x < counts.size(); x++ )
        if( !counts[x] )
            nogcol[x] = 1;

    stars.clear();
    int dy = 0;
    for( int y = 0; y < lines.size(); y++ )
    {
        string & row = lines[y];
        if( nogrow[y] )
        {
            dy += delta;
            continue;
        }
        int dx = 0;
        for( int x = 0; x < row.size(); x++ )
        {
            if( nogcol[x] )
                dx += delta;
            else if( row[x] == '#' )
                stars.push_back( Point({x+dx,y+dy}) );
        }
    }
}


int64_t part1( StringVector & lines, int delta )
{
    vector<Point> stars;
    expand(lines, delta-1, stars);

    int64_t mandist = 0;
    for( int i = 0; i < stars.size() - 1; i++ )
        for( int j = i; j < stars.size(); j++ )
            mandist += abs(stars[i].x-stars[j].x) + abs(stars[i].y-stars[j].y);

    return mandist;
}


int main( int argc, char ** argv )
{
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    StringVector lines;
    if( TEST )
    {
        istringstream data;
        data.str(test);
        string line;
        while( getline( data, line ) )
            lines.push_back( line );
    }
    else 
    {
        ifstream data("day11.txt");
        string line;
        while( getline( data, line ) )
            lines.push_back( line );
    }

    cout << "Part 1: " << part1(lines, 2) << "\n";
    if( TEST )
    {
        cout << "Test 10: " << part1(lines, 10) << "\n";
        cout << "Test 100: " << part1(lines, 100) << "\n";
    }
    cout << "Part 2: " << part1(lines, 1000000) << "\n";
}
